nodeJS/apis/RESTful_APIs.md

```diff
@@ -0,0 +1,127 @@
+
+Since we have already talked about separating the client and the server, it fulfills the first constraint of REST - the two are well-defined as the frontend and the backend. Further constraints like statelessness and caching are covered by and ensured later with ExpressJS. Our key concern at this point is the organization of endpoint URLs with respect to our resources.
+
+As mentioned, usually your backend application will need to send data to your frontend. The most popular way to do so by far is with JSON, primarily due to the ease of parsing it with JavaScript. So all we need to do is to replace our HTML and serve JSON instead. All that you have to do, thus, is to pass your information to [`res.json()`](https://expressjs.com/en/4x/api.html#res.json) instead of [`res.send()`](https://expressjs.com/en/4x/api.html#res.send) or [`res.render()`](https://expressjs.com/en/4x/api.html#res.render).
+
+### Assignment
+
+<div class="lesson-content__panel" markdown="1">
+
+1. Check out the [list of HTTP response status codes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status) listed in the MDN documentation. While you will never use some codes, you will notice a lot of useful ones.
+1. Read up on the following resources about REST and RESTful APIs.
+   - Go through the following [CodeAcademy article discussing REST](https://www.codecademy.com/article/what-is-rest). It discusses the definitions of REST and gives some good examples to explain important keywords.
+   - Check out this [StackOverflow article about RESTful API design](https://stackoverflow.blog/2020/03/02/best-practices-for-rest-api-design). If you want to code along, do remember that the `body-parser` middleware mentioned here has been incorporated into the ExpressJS package itself since 4.16.x and is no longer necessary.
+1. Code along with this tutorial on [setting up a REST API with ExpressJS](https://www.robinwieruch.de/node-express-server-rest-api/). Go through this article as thoroughly as possible since it discusses some key details like organization, middlewares, and links some important resources at the end.
+
+</div>
+
+### Knowledge check
+
+The following questions are an opportunity to reflect on key topics in this lesson. If you can't answer a question, click on it to review the material, but keep in mind you are not expected to memorize or master this knowledge.
+
+- [What is CRUD?](https://www.codecademy.com/article/what-is-crud)
+- [What are the HTTP methods and where are they used?](#crud-and-http-methods)
+- [Which HTTP method does each letter in CRUD (Create, Read, Update, Delete) correspond to?](https://stackoverflow.blog/2020/03/02/best-practices-for-rest-api-design/#h-use-nouns-instead-of-verbs-in-endpoint-paths)
+- [What part of a URL is referred to as the query parameters?](#urls)
+- [Which HTTP response status code is used to indicate the permanent relocation of a resource?](#status-codes-300---399-redirection-messages)
```

Comment:
Rails' `#parameterize` ignores the ` - ` when converting. Markdownlint's MD051 rule uses a different conversion algo but that rule's disabled in the config so you shouldn't get any errors with the fixed link fragment. 

```suggestion
- [Which HTTP response status code is used to indicate the permanent relocation of a resource?](#status-codes-300-399-redirection-messages)
```

Reply:
Yeah, markdownlint's MD051 uses the github algo for link fragments but TOP uses Rails' `#parameterize`. For things with ` - ` or `/` etc. I just throw the heading in the TOP markdown preview tool, click the heading then copy the link fragment it creates.